File: end.exs
defmodule End do
  defmodule Region do
    defstruct name: "", population: 0, south_edge: 0
  end

  def tally(lines, regions) do
    Enum.reduce(lines, regions, fn line, regions ->
      fields = String.split(line, "\t")
      {latitude, _} = Enum.at(fields, 4) |> Float.parse()
      {population, _} = Enum.at(fields, 14) |> Integer.parse()

      {regions, _} =
        Enum.map_reduce(regions, false, fn
          region, false when latitude >= region.south_edge ->
            {%{region | population: region.population + population}, true}

          region, found? ->
            {region, found?}
        end)

      regions
    end)
  end

  def main do
    regions = [
      %Region{name: "North", south_edge: 0},
      %Region{name: "South", south_edge: -90}
    ]

    [filename] = System.argv()

    File.stream!(filename)
    |> tally(regions)
    |> Enum.each(&IO.puts("#{&1.name}: #{&1.population}"))
  end
end

End.main()

File: end.rb
Region = Struct.new(:name, :population, :south_edge, keyword_init: true)

def tally(file, regions)
  file.each do |line|
    fields = line.split("\t")
    latitude = fields[4].to_f
    population = fields[14].to_i
    regions.each do |region|
      if latitude >= region.south_edge
        region.population += population
        break
      end
    end
  end
end

def main
  regions = [
    Region.new(name: "North", population: 0, south_edge: 0),
    Region.new(name: "South", population: 0, south_edge: -90),
  ]
  File.open(ARGV[0]) do |file|
    tally(file, regions)
  end
  regions.each do |region|
    puts "#{region.name}: #{region.population}"
  end
end

main
